File: sdk/c/src/providers/disposablemail_app.c
/**
 * disposablemail.app 渠道
 *
 * 纯 JSON REST API，无需认证
 * 域名: @disposablemail.dev, @mailmehere.cc
 *
 * 创建邮箱流程:
 *   POST https://disposablemail.app/api/inbox
 *   Content-Type: application/json
 *   body: {}
 *   返回: {"id":"...","address":"...@disposablemail.dev","token":"...","domain":"...","expiresAt":"...","createdAt":"..."}
 *
 * 获取邮件流程:
 *   GET https://disposablemail.app/api/inbox/emails?token={token}
 *   返回: {"emails":[...],"total":0,"inbox":{"address":"...","expiresAt":"..."}}
 *
 * token 存储策略:
 *   直接存储 API 返回的 token 字符串
 */
#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DMA_BASE "https://disposablemail.app"
#define DMA_MAX_MAILS 128

/* 公共 UA */
static const char *dma_ua =
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36";

/* ========== 创建邮箱 ========== */

/**
 * 创建 disposablemail.app 临时邮箱
 *
 * 流程:
 *   1. POST /api/inbox，body 为空 JSON 对象
 *   2. 解析响应提取 address、token
 *   3. token 直接存储 API 返回的 token 字符串
 */
tm_email_info_t *tm_provider_disposablemail_app_generate(void) {
    int timeout = tm_get_config()->timeout_secs > 0 ? tm_get_config()->timeout_secs : 15;

    /* 请求头 */
    const char *headers[] = {
        dma_ua,
        "Content-Type: application/json",
        "Accept: application/json, text/plain, */*",
        "Referer: https://disposablemail.app/",
        "Origin: https://disposablemail.app",
        NULL
    };

    /* 发送请求，body 为空 JSON 对象 */
    tm_http_response_t *resp = tm_http_request(
        TM_HTTP_POST,
        DMA_BASE "/api/inbox",
        headers,
        "{}",
        timeout
    );

    if (!resp || resp->status != 200 || !resp->body) {
        TM_LOG_ERR("[disposablemail-app] 创建收件箱请求失败, status=%ld", resp ? resp->status : 0);
        tm_http_response_free(resp);
        return NULL;
    }

    /* 解析响应 JSON */
    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);

    if (!json) {
        TM_LOG_ERR("[disposablemail-app] 解析创建收件箱响应 JSON 失败");
        return NULL;
    }

    /* 提取 address 和 token */
    const cJSON *j_address = cJSON_GetObjectItemCaseSensitive(json, "address");
    const cJSON *j_token = cJSON_GetObjectItemCaseSensitive(json, "token");

    if (!cJSON_IsString(j_address) || !j_address->valuestring || !j_address->valuestring[0]) {
        TM_LOG_ERR("[disposablemail-app] 响应中缺少 address 字段");
        cJSON_Delete(json);
        return NULL;
    }

    if (!cJSON_IsString(j_token) || !j_token->valuestring || !j_token->valuestring[0]) {
        TM_LOG_ERR("[disposablemail-app] 响应中缺少 token 字段");
        cJSON_Delete(json);
        return NULL;
    }

    /* 验证邮箱格式 */
    if (!strchr(j_address->valuestring, '@')) {
        TM_LOG_ERR("[disposablemail-app] 返回的邮箱格式无效: %s", j_address->valuestring);
        cJSON_Delete(json);
        return NULL;
    }

    /* 提取可选字段 */
    const cJSON *j_expires_at = cJSON_GetObjectItemCaseSensitive(json, "expiresAt");
    const cJSON *j_created_at = cJSON_GetObjectItemCaseSensitive(json, "createdAt");

    /* 构建结果 */
    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        cJSON_Delete(json);
        return NULL;
    }

    info->channel = CHANNEL_DISPOSABLEMAIL_APP;
    info->email = tm_strdup(j_address->valuestring);
    info->token = tm_strdup(j_token->valuestring);
    info->expires_at = 0;
    info->created_at = tm_strdup(
        (cJSON_IsString(j_created_at) && j_created_at->valuestring) ? j_created_at->valuestring : ""
    );

    cJSON_Delete(json);

    TM_LOG_DBG("[disposablemail-app] 创建邮箱成功: %s", info->email);
    return info;
}

/* ========== 获取邮件 ========== */

/**
 * 获取 disposablemail.app 邮件列表
 *
 * 流程:
 *   1. GET /api/inbox/emails?token={token}
 *   2. 解析响应中的 emails 数组并标准化
 */
tm_email_t *tm_provider_disposablemail_app_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || !email) return NULL;

    int timeout = tm_get_config()->timeout_secs > 0 ? tm_get_config()->timeout_secs : 15;

    /* 构造 URL: /api/inbox/emails?token={token} */
    size_t url_len = strlen(DMA_BASE "/api/inbox/emails?token=") + strlen(token) + 1;
    char *url = (char *)malloc(url_len);
    if (!url) return NULL;
    snprintf(url, url_len, DMA_BASE "/api/inbox/emails?token=%s", token);

    /* 请求头 */
    const char *headers[] = {
        dma_ua,
        "Accept: application/json, text/plain, */*",
        "Referer: https://disposablemail.app/",
        NULL
    };

    /* 发送请求 */
    tm_http_response_t *resp = tm_http_request(
        TM_HTTP_GET,
        url,
        headers,
        NULL,
        timeout
    );
    free(url);

    if (!resp || resp->status != 200 || !resp->body) {
        TM_LOG_ERR("[disposablemail-app] 获取邮件请求失败, status=%ld", resp ? resp->status : 0);
        tm_http_response_free(resp);
        return NULL;
    }

    /* 解析响应 JSON */
    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);

    if (!json) {
        TM_LOG_ERR("[disposablemail-app] 解析获取邮件响应 JSON 失败");
        return NULL;
    }

    /* 提取 emails 数组 */
    const cJSON *j_emails = cJSON_GetObjectItemCaseSensitive(json, "emails");
    if (!cJSON_IsArray(j_emails)) {
        /* 没有 emails 字段，视为无邮件 */
        cJSON_Delete(json);
        *count = 0;
        return NULL;
    }

    int mail_count = cJSON_GetArraySize(j_emails);
    if (mail_count <= 0) {
        cJSON_Delete(json);
        *count = 0;
        return NULL;
    }
    if (mail_count > DMA_MAX_MAILS) mail_count = DMA_MAX_MAILS;

    tm_email_t *emails = tm_emails_new(mail_count);
    if (!emails) {
        cJSON_Delete(json);
        return NULL;
    }

    int valid_count = 0;
    for (int i = 0; i < mail_count; i++) {
        const cJSON *item = cJSON_GetArrayItem(j_emails, i);
        if (!cJSON_IsObject(item)) continue;

        /* 复制邮件对象用于标准化 */
        cJSON *raw = cJSON_Duplicate(item, 1);
        if (!raw) continue;

        /* 确保 to 字段存在 */
        if (!cJSON_GetObjectItemCaseSensitive(raw, "to")) {
            cJSON_AddStringToObject(raw, "to", email);
        }

        emails[valid_count] = tm_normalize_email(raw, email);
        cJSON_Delete(raw);
        valid_count++;
    }

    cJSON_Delete(json);

    if (valid_count == 0) {
        free(emails);
        *count = 0;
        return NULL;
    }

    *count = valid_count;
    return emails;
}

File: sdk/rust/src/providers/disposablemail_app.rs
/*!
 * disposablemail.app 渠道实现
 * 网站: https://disposablemail.app
 * 域名: @disposablemail.dev, @mailmehere.cc
 *
 * 流程（纯 REST JSON API，无需认证）:
 *   1. POST https://disposablemail.app/api/inbox
 *      body: {} (空 JSON 对象)
 *      返回: {"id":"...","address":"xxx@disposablemail.dev","token":"...","domain":"disposablemail.dev","expiresAt":"...","createdAt":"..."}
 *
 *   2. GET https://disposablemail.app/api/inbox/emails?token={token}
 *      返回: {"emails":[],"total":0,"inbox":{"address":"...","expiresAt":"..."}}
 *
 * token 存储: 直接存储 API 返回的 token 字符串
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::Value;

const BASE_URL: &str = "https://disposablemail.app";

/// 创建临时邮箱
/// POST /api/inbox，将返回的 token 直接存储
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let body = serde_json::json!({});

        let resp = http_client()
            .post(format!("{}/api/inbox", BASE_URL))
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
            .header("Referer", format!("{}/", BASE_URL))
            .header("Origin", BASE_URL)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("disposablemail-app: 创建邮箱请求失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "disposablemail-app: 创建邮箱返回 HTTP {}",
                resp.status()
            ));
        }

        let data: Value = resp
            .json()
            .await
            .map_err(|e| format!("disposablemail-app: 解析创建邮箱响应失败: {}", e))?;

        let address = data
            .get("address")
            .and_then(|v| v.as_str())
            .ok_or("disposablemail-app: 响应中缺少 address 字段")?;

        if address.is_empty() || !address.contains('@') {
            return Err(format!(
                "disposablemail-app: 返回的邮箱地址无效: {}",
                address
            ));
        }

        let token = data
            .get("token")
            .and_then(|v| v.as_str())
            .ok_or("disposablemail-app: 响应中缺少 token 字段")?;

        if token.is_empty() {
            return Err("disposablemail-app: 返回的 token 为空".into());
        }

        // 解析过期时间（ISO 8601 格式转毫秒时间戳）
        let expires_at = data
            .get("expiresAt")
            .and_then(|v| v.as_str())
            .and_then(|s| chrono::DateTime::parse_from_rfc3339(s).ok())
            .map(|dt| dt.timestamp_millis());

        let created_at = data
            .get("createdAt")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        Ok(EmailInfo {
            channel: Channel::DisposablemailApp,
            email: address.to_string(),
            token: Some(token.to_string()),
            expires_at,
            created_at,
        })
    })
}

/// 获取邮件列表
/// GET /api/inbox/emails?token={token}
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let em = email.trim();
    if em.is_empty() {
        return Err("disposablemail-app: 邮箱地址为空".into());
    }

    block_on(async {
        let resp = http_client()
            .get(format!("{}/api/inbox/emails?token={}", BASE_URL, token))
            .header("Accept", "application/json")
            .header("Referer", format!("{}/", BASE_URL))
            .send()
            .await
            .map_err(|e| format!("disposablemail-app: 获取邮件列表请求失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "disposablemail-app: 获取邮件列表返回 HTTP {}",
                resp.status()
            ));
        }

        let data: Value = resp
            .json()
            .await
            .map_err(|e| format!("disposablemail-app: 解析邮件列表响应失败: {}", e))?;

        let emails = match data.get("emails").and_then(|v| v.as_array()) {
            Some(arr) => arr,
            None => return Ok(vec![]),
        };

        let mut result = Vec::with_capacity(emails.len());
        for item in emails {
            let raw = serde_json::json!({
                "id": item.get("id").unwrap_or(&Value::Null),
                "from": item.get("from").or_else(|| item.get("from_address")).or_else(|| item.get("sender")).unwrap_or(&Value::Null),
                "to": em,
                "subject": item.get("subject").unwrap_or(&Value::Null),
                "text": item.get("text").or_else(|| item.get("body")).unwrap_or(&Value::Null),
                "html": item.get("html").or_else(|| item.get("html_body")).unwrap_or(&Value::Null),
                "date": item.get("date").or_else(|| item.get("receivedAt")).or_else(|| item.get("created_at")).unwrap_or(&Value::Null),
                "isRead": item.get("isRead").unwrap_or(&Value::Bool(false)),
                "attachments": item.get("attachments").unwrap_or(&Value::Array(vec![])),
            });
            result.push(normalize_email(&raw, em));
        }

        Ok(result)
    })
}
